pkg/leader/leaderelector.go

```diff
@@ -51,22 +52,32 @@ func (f *DummyElector) RunWithLeader(ctx context.Context, run func() error) erro
 var _ ElectorInterface = &Elector{}
 
 type Elector struct {
-	log           logrus.FieldLogger
-	config        Config
-	kube          *kubernetes.Clientset
-	isLeader      bool
-	configMapName string
+	log      logrus.FieldLogger
+	mu       *sync.RWMutex
+	config   Config
+	kube     *kubernetes.Clientset
+	lockName string
+	isLeader bool
 }
 
-func NewElector(kubeClient *kubernetes.Clientset, config Config, configMapName string, logger logrus.FieldLogger) *Elector {
-	logger = logger.WithField("configMap", configMapName)
-	return &Elector{log: logger, config: config, kube: kubeClient, configMapName: configMapName, isLeader: false}
+func NewElector(kubeClient *kubernetes.Clientset, config Config, lockName string, logger logrus.FieldLogger) *Elector {
+	logger = logger.WithField("lock", lockName)
+	mu := &sync.RWMutex{}
+	return &Elector{log: logger, config: config, kube: kubeClient, lockName: lockName, isLeader: false, mu: mu}
 }
 
 func (l *Elector) IsLeader() bool {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
 	return l.isLeader
 }
 
+func (l *Elector) setLeader(status bool) {
+	l.mu.Lock()
```

Comment:
same question , why we need lock? setting boolean is thread safe so why it is needed?

Reply:
I agree, lock maybe solve some races on a local pod but you will still have a race between different pods, and in wort case one of them will fail to run. so it's really depend on what you are trying to solve. 
IMO the lock doesn't avoid any issues, in worst case you will get the right reply in the next iteration.